src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs

```diff
@@ -450,8 +450,17 @@ private BoundLambda ReallyBind(NamedTypeSymbol delegateType)
                 cacheKey.ParameterTypes,
                 cacheKey.ParameterRefKinds,
                 refKind,
-                returnType);
+                returnType,
+                diagnostics);
             lambdaBodyBinder = new ExecutableCodeBinder(_unboundLambda.Syntax, lambdaSymbol, ParameterBinder(lambdaSymbol, binder));
+
+            if (lambdaSymbol.RefKind == CodeAnalysis.RefKind.RefReadOnly)
+            {
+                binder.Compilation.EnsureIsReadOnlyAttributeExists(diagnostics, lambdaSymbol.DiagnosticLocation, modifyCompilationForRefReadOnly: false);
+            }
+
+            ParameterHelpers.EnsureIsReadOnlyAttributeExists(lambdaSymbol.Parameters, diagnostics, modifyCompilationForRefReadOnly: false);
```

Comment:
For lambdas and local functions, we do that two times:
1) Here, where we want to generate diagnostics for each symbol, but not actually modify the compilation (because they might not belong to the current one). That has to be done here rather than the constructor because of circular lookups. The compiler can crash with a SOE trying to bing all types.
2) In the local rewriter, where we will modify the current module builder, and note the need for the attribute, but we will not generate errors, because they're are already generated here.